### ISSUE TITLE
Trim multiple empty lines

### ISSUE THREAD

**Issue by csleead** [editorconfig/editorconfig#336]
One empty line is useful for separating code of different logic but having multiple empty lines is ugly.
Can we have an option to trim multiple empty lines into one?

**Comment by kleinech** [comment]
Here is an example of IntelliJ

![image](https://user-images.githubusercontent.com/18282351/36966571-784ff15c-205d-11e8-8b6e-1e1f73859239.png)

**Comment by xuhdev** [comment]
@editorconfig/developers  What do you think? I personally think such an option may be useful for some kinds of projects.

**Comment by sindresorhus** [comment]
👎 It's not really possible to do this reliably without parsing the code. How would EditorConfig know if the blank lines are inside a comment or multi-line string?

**Comment by kleinech** [comment]
Could it be a candidate for a `Domain-specific property`.

**Comment by xuhdev** [comment]
I agree with @sindresorhus , and personally I think programmers intend to have more than 1 consecutive blank lines purposely from time to time. Now I'm also curious: what language do you use IntelliJ for? Maybe that language has a specific feature.

**Comment by kleinech** [comment]
The language is Java.

But it is more like a style guide thing: multiple blank lines often occur by accident and causes errors:
```
[ERROR] src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/boundarydefinitions/PropertyConstraintsResource.java:[91] (whitespace) EmptyLineSeparator: There is more than 1 empty line one after another.
```
https://eslint.org/docs/rules/no-multiple-empty-lines

**Comment by koppor** [comment]
I would use this setting independently of the context. It just disallows double empty lines everywhere. In case a user wants to have double lines in a certain file, he has to configure editorconfig differently for that file.

Similar examle: If editorconfig is configured using `space` as indent. Now, I write a bash-script creating a file using `tab` as indent by enclusing it using `cat <<<EOF`... Such a block is not recognized by editorconfig, but still would change the file in a non-wanted way. Then, as user I have to adapt my editorconfig to this file so that it does not treat this file.

**Comment by jedmao** [comment]
@sindresorhus' [comment above](https://github.com/editorconfig/editorconfig/issues/336#issuecomment-370426806) trumps any possible request to implement this into EditorConfig. It doesn't matter how handy the setting would be if you can't rely on it not breaking things, which is exactly the case here. You would have to have a language-specific linter or fixer like `eslint fix` to do a task like this (something that understands the syntax of a comment in that language), which is outside the scope of EditorConfig.

I think we can close this issue now.

**Comment by jedmao** [comment]
@hax the bottom line is that for EditorConfig to support this rule it would have to be aware of a specific language syntax in order to prevent it from breaking something like this in JavaScript:

```js
const text = `foo


bar`;
```
The above is a perfectly valid multi-line string, but we only know this because we understand how JavaScript handles multi-lines. EditorConfig does not understand language-specific constructs, so it remains unable to do stuff like this. Why? Because it would reduce the above code into:
```js
const text = `foo
bar`;
```
The result of which is `foo\nbar`, totally different from `foo\n\n\nbar` above.

Now, ESLint has a rule called [`no-multiple-empty-lines`](https://eslint.org/docs/rules/no-multiple-empty-lines) that would give you what you want in a language-specific way. That would be the correct way to go about it in the case of the example above and it shouldn't break the original content of the multi-line string.

**Comment by hax** [comment]
@jedmao 

I hope I could make it more clear. So let me try again.

---

First, "empty lines in multiline comments" is not a problem. For example, ESLint `no-multiple-empty-lines` rule you mentioned also apply to multiline comments.

Two, many languages do not have "empty lines in multiline string" problem because they do not have multiline strings, or they do not allow empty lines in multiline strings. For example many C-family languages require `\` at the end of each line of multiline strings.

Three, for those allow "empty lines in multiline string", some team adopt the coding style that also restrict it and force the programmers use escape (like `\n`), placeholder (like `$NEWLINE`) or other form (string concat, etc.) instead of rely on error-prone multiple empty lines. In fact, it's really rare to need many multiple empty lines in string content, most time they are typos. So in my opinion, ESLint `no-multiple-empty-lines` rule should have a extra option to also apply to template string, I'll raise an issue in ESLint repo later.

Four, if some do not want to apply the restriction to "empty lines in multiline string", they always can  disable the property in editorconfig, and use language-specific linters like ESLint.

---

Especially, we should notice that "empty lines in multiline string" problem is a more general problem, that is: Does the coding style apply to the content of multiline strings? 

For example, "tabs/spaces in multiline string", how `indent_style`, `indent_size` could/should work? FYI, ESLint `indent` rule won't complain for inconsistency in template string -- aka, language-specific!  

This also apply to `end_of_line`, `trim_trailing_whitespace`.

We should think `max_empty_lines` (let's temporarily use this name) in a consistent way.

---

The real problem is what should editor/IDE/tool do when meet extra empty lines. At least three options:

1. Show some warning but not trim it automatically (more like a linter).
2. Do not allow enter extra empty lines, and/or when a line is modified, collapse the adjacent empty lines, but do not touch non-modified lines.
3. Silently trim extra empty lines on save.

It seems you against `max_empty_lines` because you only consider the third option. But we really have other options.

Actually this is also a more general problem of editorconfig, and it's the root of some issues in the repo.

Most plugin adopt the second option for `indent_style` (only convert tab/spaces in modified lines, do not touch non-modified lines)

Most plugin adopt the third option for `trim_trailing_whitespace` and `insert_final_newline` (silently trim/insert on save)

CLI tools like "eclint" and "lintspace" adopt the first option for all properties when check, and adopt the third option for all when autofix. (which means they will already suffer similar "change content of multiline strings" problem on `indent_style` and other properties when autofix.)

So I think the real problem is we need a guide for precise and consistent behavior across all properties for the authors of plugins.

Anyway, it's safe  to support `max_empty_lines` using the option 2.

---

Hope I make it clear now. Can we reopen this issue? Thank you.

**Comment by thebarty** [comment]
I like this idea a lot!! +1 to have it implemented.

Personally I would simply disallow empty lines… I find the code more readable without them and use comments to separete logical parts

**Comment by jedmao** [comment]
The issue with all three of your options is that there's no way to detect an empty line w/o knowledge of the language in question (parsing it into some kind of AST), so you still haven't addressed this issue.

@thebarty you can disallow empty lines with a linter.

**Comment by jedmao** [comment]
@Kirlac, I'm curious what language you're writing in? And do you use a linter? If not, why? Linters are aware of AST. EditorConfig is not. You might be interested in [eclint](https://github.com/jedmao/eclint).

You bring up a good argument with multi-line indentation. There may very well be EditorConfig extensions out there in the wild that are doing something about the indentation in these cases, but I would strongly advise against it. I personally think it's the job of any EditorConfig extension to set the "editor's configuration" settings and defer the behavior of enforcing it to the editor itself. In the case of vscode, the EditorConfig extension just sets the indentation setting and calls it a day.

Some EditorConfig rules are nouns (passive). Some are verbs (active). The passive ones should remain passive and the active ones should defer the action to built-in editor features. In the case of vscode:

```json
{
    "files.insertFinalNewline": true,
    "files.trimTrailingWhitespace": true,
}
```
No additional work required.

See also: https://github.com/editorconfig/editorconfig-vscode/issues/85

**Comment by Kirlac** [comment]
@jedmao Thanks for your response and explanation. I wasn't sure how that was being handled - it makes sense to just set this in the editor itself - the last thing you want is the application and editorconfig plugin fighting over how to do things!

To answer your question, I'm using a variety of languages - c#, typescript, css, html, json, sql, cypher to name my common day-to-day ones for work - plus whatever languages are being used by the pet projects i pick up in my spare time, or whatever bat/sh/ps1 script or xml/ini based config I need to include. Obviously not all of those require this functionality, but my point was that it would be nice to have a language independent option available via an editorconfig rule.

To clarify, I wasn't advocating against linting - linters are awesome. I was simply suggesting that linting shouldn't need to factor into this feature and would be drastically over-engineering what (I assume) would otherwise be a simple regex to replace multiple empty lines with a single one, and that there is little harm in providing an "always trim to 1 consecutive empty line" option regardless of context and language for those who require it. Those who don't would simply not enable the rule and handle it whatever way works for them eg. by configuring their linter to handle it, or by simply not caring about the extra lines.

Again, I'm not saying I disagree with you. I can certainly see your reasoning on this, and it's definitely valid. I'm just trying to give an alternate perspective based on my experience, which may not have been considered/understood in the context of this feature request.

**Comment by jedmao** [comment]
I 100% understand and respect your desire to have this rule. I've been there myself. At first, it seems easy and harmless, just as you said. You might be aware of the rule now when you implement it in your project, but people move around a lot, from company to company. One day, you'll be gone and someone else will be there and will have no idea why their code is doing unexpected things. Personally, I don't want to be responsible for introducing such a shady rule.

That said, I started the [eclint](https://github.com/jedmao/eclint) project because I was naive. I never expected it to be as popular as it is now and I'm ashamed of the project, really, because it violates the tenets in which I now hold true. Hell, I used to think it was possible to fix indentation w/o AST, but I was foolish. I spent some serious time in the trenches, convinced I could solve this problem (w/o AST). Now I'm convinced it's impossible. I didn't "give up". I just realized it was literally impossible, something I'm rarely comfortable saying about code.

My philosophy behind EditorConfig is much more simple now than it ever was before. Let it stick to editor configuration settings and no extra nonsense. I've done all the extra nonsense in the Visual Studio extension and it just over-complicated everything. I don't even want to touch that project again. Just let the editors do what they do best.

**Comment by Kirlac** [comment]
Damn I can't argue with any of that. You're absolutely right. I'm looking for an easy solution to stop my juniors opening PRs that are half made up of newlines spammed around their code. It seems like this would be a harmless fix, but I didn't think about what would happen if I handed over the reigns to someone else. And to top it off, it shouldn't be editorconfigs job to enforce that. I guess a part of me (and I assume others) view this as an extension to the editor's configuration, rather than a simple and unified way of controlling it, which I think is a fundamental misunderstanding on my part.

I completely get what you mean. I think it's that shame of looking back and seeing the violation of those tenets that drives us to be better, and I certainly wouldn't ask you to compromise on that. It sounds like you have far more insight into the mechanics of this than I do, and impossible is certainly a strong word to use there. Shortcuts in code are very rarely the right answer. If it's worth doing it should be done right and as you've pointed out there are tools already available which have this done right.

I withdraw my previous comment. This isn't your issue to fix. Thanks for the discussion, it's been enlightening
